In [1]:
import numpy as np
from scipy.special import comb
import math

# $Markov\,Chains$

By **ShineHale**(yunsong Yang) U202115980

This is the second experiment for the random process, and it is about the markov chains, which is one of the most important models in the process of exploring the bound of the world.

## $Task \, 1$

Consider a markov chain with the following transition probability matrix

![](./img/mc0.png)

- (a) Decompose the state space into equivalance classes.
- (b) Identify the close equivalence clases and find their respective periods.
- (c) State the transition probability matrix in a canonical form.

### $Quesiton \,1$
we can easily plot the state transmition figure like the follow:

![](./img/mc1.jpg)


then from the plot, we can have the conclusion that, there are $4$ equivalance classes. 
- $\{9\}$ 
- $\{4\}$
- $\{1, 7, 10\}$
- $\{2, 3, 5, 8\}$ 


### $Question\, 2$
For each the closed equivalance classes, the period of them each one is:

-  $pid(\{9\}) = 1$
-  $pid(\{4\}) = 1$
-  $pid(\{1, 7, 10\}) = 3$
-  $pid(\{2, 3, 5, 8\}) = 1$

### $Question\, 3$
we can state the transition probability matrix in a canonical form as follow:

$$
P = \left[
        \begin{matrix}
            0 & 0 & 0 & 0 & 0 & 0 & 1.0 & 0 & 0 & 0 \\
            0 & 0.3 & 0.3 & 0.1 & 0.3 & 0 & 0 & 0 & 0 & 0 \\
            0 & 0 & 0.6 & 0 & 0& 0 & 0 & 0.4 & 0 & 0 \\
            0 & 0 & 0 & 1 & 0 & 0 & 0 & 0 & 0 & 0 \\
            0 & 0.4 & 0 & 0 & 0 & 0.3 & 0 & 0.3 & 0 & 0 \\ 
            0 & 0 & 0 & 0.9 & 0 & 0 & 0.1 & 0 & 0 & 0 \\ 
            0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 1.0 \\
            0 & 0.8 & 0 & 0 & 0 & 0 & 0 & 0 & 0.2 & 0 \\
            0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 1.0 & 0 \\
            1.0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 
        \end{matrix}
    \right]
$$

## $Task \, 2$

A fugitive moves about three cities, Houston, Galveston, ana Sugarianu, to escape a bounty hunter. Initially, the fugitive is in Houston and the bounty hunter is in Galveston. The fugitive and bounty hunter move independently of each other and each follows the respective transition probability matrix of a Markov chain:



$$
P_{fugitive} = \begin{matrix}
        Houston \\
        Galveston \\
        Sugarland  \\
    \end{matrix}
    \left[\begin{matrix}
        0 & 0.4 & 0.6 \\
        0.7 & 0 & 0.3 \\
        0.1 & 0.9 & 0 \\
    \end{matrix}\right] \,
P_{bounty \, hunter} = \begin{matrix}
        Houston \\
        Galveston \\
        Sugarland  \\
    \end{matrix}
    \left[\begin{matrix}
        0.3 & 0.2 & 0.5 \\
        0.8 & 0.1 & 0.1 \\
        0.5 & 0.2 & 0.3 \\
    \end{matrix}\right]
$$

We assume that each party makes the move (transition) at the end of each day. When the two are in the same city on a given day there is a 50 percent chance the fugitive will be captured. What is the expected time for the bounty hunter to catch the fugitive?


In [2]:
p_fu = np.array([[0, 0.4, 0.6],
                 [0.7, 0, 0.3],
                 [0.1, 0.9, 0]])
p_bh = np.array([[0.3, 0.2, 0.5],
                 [0.8, 0.1, 0.1],
                 [0.5, 0.2, 0.3]])
p_init_fu = np.array([1, 0, 0])
p_init_bh = np.array([0, 0, 1])

In [3]:
def generate_outcome():
    days = 0
    pos_fu, pos_bh = 0, 2
    while True:
        dir_fu = np.random.rand()
        for k in range(3):
            if p_fu[pos_fu][k] > dir_fu:
                pos_fu = k
                break
            else:
                dir_fu -= p_fu[pos_fu][k]
        dir_bh = np.random.rand()
        for k in range(3):
            if p_bh[pos_bh][k] > dir_bh:
                pos_bh = k
                break
            else:
                dir_bh -= p_bh[pos_bh][k]
        if pos_bh == pos_fu:
            case = np.random.rand()
            if (case < 0.5):
                break
        days += 1
    return days

In [4]:
times = 100000
total_days = 0
for _ in range(times):
    total_days += generate_outcome()
print("the expected time for the bounty hunter to catch the fuguitive is %.4f days" %(total_days / times))

the expected time for the bounty hunter to catch the fuguitive is 5.6762 days


## $Task \, 3$

Consider Problem 5 again. Assume that emergency arrivals follow a Poisson process with a rate of one arrival per day and the number of discharges per day, given that there are $i$ patients at the start of the day, follows a binomial distribution with parameters $(i, p)$, where $p = 0.5$. For the first patient admission policies, find 
- (a) the one-step transition probability matrices; 
- (b) the long-run expected occupancies per day; 
- (c) the long-run expected number of emergency patients who are turned away per day. Under the current assumption of binomial discharges, what type of length-of-stay distributions are we assuming?

### $Question \, 1$
we consider the policy of only admitting emergency arrivals to the ward, denote $\{X_t|t = 0, 1, \cdots\}$ as the number of the patients in the ward in the $t$ day, then we will have 

$$P\{X_{i + 1} = k| X_i = m\} = \sum_{n = 0}^{m} \binom{m}{n}p^n(1-p)^{(n-m)}  P\{N(1) = k - m + n\}$$
subsitute the $P\{N(1) = k - m + n\}$, we will have
$$P\{X_{i + 1} = k| X_i = m\} = \sum_{n = 0}^{m} \binom{m}{n}p^n(1-p)^{(n-m)}  \frac{\lambda^{k - m + n}}{(k - m +n)!}exp(-\lambda)$$
where $k - m + n >= 0, \lambda = 1, p = 0.5$

so we can have the one-step transition matrix as follow.

$$
$$

In [5]:
P = np.zeros((4, 4))
p = 0.5
lambd = 1
for m in range(4):
    for k in range(3):
        val = 0
        for n in range(0, m + 1, 1):
            if k - m + n < 0:
                continue
            val += comb(m, n) * p**n * (1-p)**(m - n)  * np.exp(-lambd) * lambd**(k - m + n) / math.factorial(k - m + n)
            P[m][k] = val
    P[m][3] = 1 - sum(P[m])
print(P)

[[0.36787944 0.36787944 0.18393972 0.0803014 ]
 [0.18393972 0.36787944 0.27590958 0.17227126]
 [0.09196986 0.27590958 0.32189451 0.31022605]
 [0.04598493 0.18393972 0.29890205 0.4711733 ]]


### $Question\, 2$

In order to find the long-run expected occupancies per day, we need to find the long-run expected number of the patients of per day in the ward. we can assume it has the limit distribution(that is true), and iterate the matrix over and over again. finally, we will have the limit distribution, then have the expected number of the patients.

In [6]:
times = 30
P_iter = P
for _ in range(times):
    P_iter = np.dot(P_iter, P_iter)
print(P_iter)

[[0.14594531 0.29006995 0.28193685 0.28204794]
 [0.14594531 0.29006995 0.28193685 0.28204794]
 [0.14594531 0.29006995 0.28193685 0.28204794]
 [0.14594531 0.29006995 0.28193685 0.28204794]]


after make the operations about 30 times, we can find that the matrix is fixed, what a amazing outcome, it means $P_{limit} = P^{2^{30}}$, and no matter how you times itself, its value is stick!, so we have the limit distribution $[0.1459, 0.2901, 0.2819, 0.2820]$, then we can have the expected number of the patients and the long-run expected occupancies per day. 

In [7]:
Ex_patients = np.sum(P_iter[0] * np.array([0, 1, 2, 3]))
Ex_rate = Ex_patients / 3
print("the long-run expected occupancies per day is %.4f%%" %(Ex_rate * 100))

the long-run expected occupancies per day is 56.6696%


### $Question \, 3$

Under the assumption of binomial discharges, we are assuming a fixed length-of-stay distribution. This is because the binomial distribution assumes that each patient has a fixed probability of being discharged on any given day, regardless of how long they have been in the hospital. Therefore, the length of stay for each patient is assumed to be independent and identically distributed (iid) with a fixed probability of discharge on each day.

## $Task \, 4$

An input conveyor panses by :a work station every twelve minutes with an item to be processed by the operator at the work station. The amount of work for each item differs depending on the instructions stated on the tag of each item. There are three types of items: 2/7 of them require six-minutes of work, 4/7 of them require twelve minutes of work, and 1/7 of them require eighteen-minutes of work. The worker at the work station works on one item at a time.The station has a storage space for holding one unprocessed item. If an item is completed in the middle of the twelve-minute interval, it is immediately loaded on an output conveyor and the item at the storage space, if any, is moved to the work station for immedate processing. At the end of a twelve-minute interval, the worker loads the finished item, if any, on the output conveyor and moves the waiting item at the storage space, if any, to the work station for immediate processing. In  addition, the ariving item from the input conveyor is moved to the work station or the storage space. depending on the space available at either place. Assume that loading and unloading times are negligible. When an item passes by the work station and the work station does not have a place to store it, the item leaves the system unprocessed. The figure above depicts the complete system. Define the work load at the work station as the amount of work remaining (in minutes of work yet to be completed) right before the conveyor passes by the work station.

- (a) Compute the average amount of work load at the work station. 
- (b) Compute the fraction of items left in the work station unprocessed.

### $Question \, 1$

In case of the calculation of the probabilities, we prefer to simulate the process and using the long-run process for solving the problem.


In [8]:
def generate_item():
    pro = np.random.rand()
    costs_time = 0
    if pro < 2.0 / 7:
        costs_time = 6
    elif pro < 6.0 / 7:
        costs_time = 12
    else:
        costs_time = 18
    return costs_time

In [9]:
max_times = 5000000
empty = 0
tot_amount = 0
tot_work_load = 0

left_time = 0
for _ in range(max_times + 1):
    cost = generate_item()

    count = 0
    time = 12
    if left_time <= time and left_time > 0:
        count += 1
        time -= left_time
        left_time = empty
        empty = 0
    else:
        if left_time > 0:
            left_time -= time
        time = 0
    if left_time <= time and left_time > 0:
        count += 1
        time -= left_time
        left_time = 0
    else:
        if left_time > 0:
            left_time -= time
        time = 0
    
    tot_work_load += left_time + empty
    tot_amount += count
    if left_time == 0:
        left_time = cost
    elif empty == 0:
        empty = cost
    
print("the average work load of the working staion is %.2f minutes" %(tot_work_load / max_times))

the average work load of the working staion is 3.37 minutes


### $Question \, 2$

In order to compute the fraction of items left in the work station unprocessed,  we only need to compute the number of the processed items completely, and use the formula:

$$
P = 1 - \frac{tot_{amount}}{tot_{times}}
$$


In [10]:
max_times = 5000000
empty = 0
tot_amount = 0
tot_work_load = 0

left_time = 0
for _ in range(max_times + 1):
    cost = generate_item()

    count = 0
    time = 12
    if left_time <= time and left_time > 0:
        count += 1
        time -= left_time
        left_time = empty
        empty = 0
    else:
        if left_time > 0:
            left_time -= time
        time = 0
    if left_time <= time and left_time > 0:
        count += 1
        time -= left_time
        left_time = 0
    else:
        if left_time > 0:
            left_time -= time
        time = 0
    
    tot_work_load += left_time + empty
    tot_amount += count
    if left_time == 0:
        left_time = cost
    elif empty == 0:
        empty = cost
    
print("the fraction of items left in the work station unprocessed is %.2f%%" %((1 - tot_amount / max_times) * 100))

the fraction of items left in the work station unprocessed is 1.42%


## $Summary$

From this experiment, I have a better understanding about the markov chain like limit distribution, states classification and apply them into the specific question, for solving the specific problem. What a wonderful journey! 